{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div align=\"right\" vertical-align=\"middle\" style=\"border: 2px solid;border-radius: 5px;background-color:lightgrey;padding:5px;padding-right:20px;padding-left:10px;\">\n",
    "        <a style=\"color:black;text-decoration:none;\" title=\"Home\" href=\"../index.ipynb\">\n",
    "            <img src=\"../../css/iconmonstr-christmas-house-icon.svg\" height = \"30\" width = \"30\" style=\"display:inline\">\n",
    "        </a>\n",
    "        &nbsp;\n",
    "        <b>|</b>\n",
    "        &nbsp;\n",
    "        <a style=\"color:black;text-decoration:none;\" title=\"Build\" href=\"../build_docs/build.ipynb\">\n",
    "            <img src=\"../../css/iconmonstr-puzzle-icon.svg\" height = \"30\" width = \"30\" style=\"display:inline\">\n",
    "        </a>\n",
    "        <a style=\"color:black;text-decoration:none;\" title=\"Assemble\" href=\"assemble.ipynb\">\n",
    "            <img src=\"../../css/iconmonstr-puzzle-17-icon.svg\" height = \"30\" width = \"30\" style=\"display:inline\">\n",
    "        </a>\n",
    "        <a style=\"color:black;text-decoration:none;\" title=\"Query\" href=\"../query_docs/query.ipynb\">\n",
    "            <img src=\"../../css/iconmonstr-flask-3-icon.svg\" height = \"30\" width = \"30\" style=\"display:inline\">\n",
    "        </a>\n",
    "</div>\n",
    "\n",
    "# TUTORIAL: Assemble an EGRIN 2.0 Ensemble\n",
    "\n",
    "####*Important!!! This tutorial assumes you have access to a complete cMonkey2 ensemble.*\n",
    "\n",
    "### In a nutshell\n",
    "\n",
    "The `ASSEMBLE` scripts transfer and compile individual [cMonkey2](https://github.com/baliga-lab/cmonkey2) SQLite databases into an integrated MongoDB database. \n",
    "\n",
    "In addition, they perform several post-processing steps, including: detection of gene regulatory elements (GREs) by comparing individual bicluster motifs with [`TOMTOM`](http://meme.ebi.edu.au/meme/doc/tomtom.html) and clustering with [MCL](http://micans.org/mcl/), genome-wide scanning of motifs with FIMO, and detection of co-regulated modules or **corems** using link-community detection.\n",
    "\n",
    "### Requirements\n",
    "\n",
    "- MongoDB >= 2.4.9\n",
    "- compiled C++ scripts for corem detection, available [here](cpp_scripts.ipynb)\n",
    "\n",
    "IMPORTANT: This tutorial currently assumes that `TOMTOM`, `MCL` and `FIMO` have already been run.\n",
    "\n",
    "A single GRE definition file is read from, eg:\n",
    "    \n",
    "    /ensemble-head-dir\n",
    "        /out.mot_metaclustering.txt.I45.txt\n",
    "        \n",
    "FIMO scans are read from each run sub-directory, eg:\n",
    "    \n",
    "    /ensemble-head-dir\n",
    "        /org-out-xxx\n",
    "            /fimo-outs\n",
    "                /fimo-out-xxxx.bz2\n",
    "\n",
    "Optional:\n",
    "- `row_annot`: tab-delimited row (gene) annotations. Will be downloaded from MicrobesOnline automatically using --ncbi_code if undefined \n",
    "- `col_annot`: tab-delimited column (condition) annotations.\n",
    "\n",
    "The format for these files will be described in detail below.\n",
    "\n",
    "Additionally, the Python modules described on the [Home page](../index.ipynb) are required to run these scripts.\n",
    "\n",
    "### Scripts\n",
    "\n",
    "- `assembler.py`: The control function for `ASSEMBLE` scripts.\n",
    "- `makeCorems.py`: Identifies corems using C++ scripts compiled above\n",
    "- `resample_QSub.py`: Generates QSub script for submission of resamples to cluster\n",
    "- `sql2mongoDB.py`: Merges individual cMonkey SQLite dbs and post-processing data into MongoDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: assembler.py [-h] --organism ORGANISM --ratios RATIOS --targetdir\n",
      "                    TARGETDIR --ncbi_code NCBI_CODE [--cores CORES]\n",
      "                    [--ensembledir ENSEMBLEDIR] [--col_annot COL_ANNOT]\n",
      "                    [--host HOST] [--port PORT] [--prefix PREFIX]\n",
      "                    [--row_annot ROW_ANNOT]\n",
      "                    [--row_annot_matchCol ROW_ANNOT_MATCHCOL]\n",
      "                    [--gre2motif GRE2MOTIF] [--db DB]\n",
      "                    [--genome_annot GENOME_ANNOT]\n",
      "                    [--backbone_pval BACKBONE_PVAL]\n",
      "                    [--link_comm_score LINK_COMM_SCORE]\n",
      "                    [--link_comm_increment LINK_COMM_INCREMENT]\n",
      "                    [--link_comm_density_score LINK_COMM_DENSITY_SCORE]\n",
      "                    [--corem_size_threshold COREM_SIZE_THRESHOLD]\n",
      "                    [--n_resamples N_RESAMPLES] [--cluster CLUSTER]\n",
      "                    [--finish_only FINISH_ONLY] [--user USER]\n",
      "\n",
      "assemble.py - prepare cluster runs\n",
      "\n",
      "optional arguments:\n",
      "  -h, --help            show this help message and exit\n",
      "  --organism ORGANISM   3 letter organism code\n",
      "  --ratios RATIOS       Path to ratios file. Should be 'raw' (normalized)\n",
      "                        ratios, not the standardized ratios used by cMonkey\n",
      "  --targetdir TARGETDIR\n",
      "                        Storage path for MongoDB and corem data\n",
      "  --ncbi_code NCBI_CODE\n",
      "                        NCBI organism code\n",
      "  --cores CORES         Number local cores to use for corem C++ scripts\n",
      "  --ensembledir ENSEMBLEDIR\n",
      "                        Path to ensemble runs. Default: cwd\n",
      "  --col_annot COL_ANNOT\n",
      "                        Tab-delimited file with experiment annotations\n",
      "  --host HOST           MongoDB host. Default 'localhost'\n",
      "  --port PORT           MongoDB port\n",
      "  --prefix PREFIX       Ensemble run prefix. Default: *organism*-out-\n",
      "  --row_annot ROW_ANNOT\n",
      "                        Optional row (gene) annotation tab-delimited file. If\n",
      "                        not specified, annotations will be downloaded from\n",
      "                        MicrobesOnline using --ncbi_code.\n",
      "  --row_annot_matchCol ROW_ANNOT_MATCHCOL\n",
      "                        Name of column in row_annot that matches row names in\n",
      "                        ratios file.\n",
      "  --gre2motif GRE2MOTIF\n",
      "                        Motif->GRE clustering file\n",
      "  --db DB               Optional ensemble MongoDB database name\n",
      "  --genome_annot GENOME_ANNOT\n",
      "                        Optional genome annotation file. Automatically\n",
      "                        downloaded from MicrobesOnline using --ncbi_code\n",
      "  --backbone_pval BACKBONE_PVAL\n",
      "                        Significance pvalue for gene-gene backbone. Default =\n",
      "                        0.05.\n",
      "  --link_comm_score LINK_COMM_SCORE\n",
      "                        Scoring metric for link communities\n",
      "  --link_comm_increment LINK_COMM_INCREMENT\n",
      "                        Height increment for cutting agglomerative clustering\n",
      "                        of link communities\n",
      "  --link_comm_density_score LINK_COMM_DENSITY_SCORE\n",
      "                        Density score for evaluating link communities\n",
      "  --corem_size_threshold COREM_SIZE_THRESHOLD\n",
      "                        Defines minimum corem size. Default = 3.\n",
      "  --n_resamples N_RESAMPLES\n",
      "                        Number resamples to compute for corem condition\n",
      "                        assignment. Default = 10,000\n",
      "  --cluster CLUSTER     Run re-samples on cluster? Boolean.\n",
      "  --finish_only FINISH_ONLY\n",
      "                        Finish corems only. In case session gets dropped\n",
      "  --user USER           Cluster user name\n"
     ]
    }
   ],
   "source": [
    "%run ../../assemble/assembler.py -h"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### `ASSEMBLE` an EGRIN 2.0 ensemble\n",
    "\n",
    "In this tutorial we will see how you would `ASSEMBLE` an *Escherichia coli* EGRIN 2.0 ensemble using several example files and a couple of cMonkey2 runs, which we provide [here](./static/example_files/).\n",
    "\n",
    "### STEP 1: Generate optional input files\n",
    "\n",
    "First, let's explore the optional annotation files. Providing annotations for genes and conditions is a great way to enrich your analysis of the ensemble. You can get a better idea for the utility of these metainformation by following the [advanced mining tutorial](../query_docs/advanced_mining.ipynb)\n",
    "\n",
    "#### `row_annot`\n",
    "\n",
    "As noted above, the `row_annot` file will be downloaded automatically from MicrobesOnline if a custom annotation is not provided. If you provide your own `row_annot` file, however, you will also need to specificy `--row_annot_matchCol`, which is the name of the column in your annotation file that matches the gene name used by cMonkey2 (i.e. the row names in your ratios file). \n",
    "\n",
    "The row annotation file should look like the annotation file supplied by MicrobesOnline, where each row specifies a gene and each of the columns specifies some information about that gene. Again, you must ensure that at least one of the columns contains gene names that match the gene names in the ratios file used by cMonkey2, in the case of MicrobesOnline, it is the `sysName` column.\n",
    "\n",
    "Here is an example annotation file for *E. coli* direct from MicrobesOnline, the file itself is available [here](./static/example_files/511145_geneInfo.tab).\n",
    "\n",
    "<img src=\"./static/row_annot.png\" height = \"75%\" width = \"75%\" style=\"display:inline\">\n",
    "\n",
    "#### `col_annot`\n",
    "\n",
    "The `col_annot` file provides metainformation about each experiment. Like the `row_annot` file, these annotations are optional, but they can be valuable for making sense of ensemble predicitions. \n",
    "\n",
    "Please note that the file format is different here. Each row contains a particular experimental meta-annotation followed by several required descriptions: (1) `experiment_name`, (2) `feature_name`, (3) `value`, (4) `feature_units`, (5) `feature_type`. The `experiment_name` column should match the experiment name in the `ratios` file.\n",
    "\n",
    "The `col_annot` file should look like the tab-delimited file depicted below. You can download an *E. coli* `col_annot` file to use as a template [here](./static/example_files/E_coli_v4_Build_6.experiment_feature_descriptions.tsv)\n",
    "\n",
    "<img src=\"./static/experiment_annotation.png\" height = \"75%\" width = \"75%\" style=\"display:inline\">\n",
    "\n",
    "### STEP 2: Run `assembler.py`\n",
    "\n",
    "<u>Important!!! TOMTOM, MCL, and FIMO should be run prior to assembly. Otherwise the ensemble will not contain GREs or motif scans.</u>\n",
    "\n",
    "To run the `assembler`, you must supply several files as well as specify where you would like to host the MongoDB database. If you do not supply a `host` for the MongoDB databse, it will be stored on `locahost`. At a minimum, you should supply:\n",
    "\n",
    "    --organism, 3 letter organism code\n",
    "    --ratios, Path to ratios file. Should be 'raw' (normalized) ratios, not the standardized ratios used by cMonkey\n",
    "    --targetdir, Storage path for MongoDB and corem data\n",
    "    --ncbi_code, NCBI organism code\n",
    "    --ensembledir, Path to ensemble runs. Default: cwd\n",
    "\n",
    "Assuming your terminal current working directory is within the *E. coli* ensemble, with the cMonkey2 runs located in the subfolder `./eco-ens-m3d/`, you can call the `assembler` as follows: \n",
    "\n",
    "    python assembler.py --organism eco --ratios ./ratios_eco_m3d.tsv.gz --targetdir ./ --ncbi_code 511145 --ensembledir ./eco-ens-m3d/ --col_annot ./E_coli_v4_Build_6.experiment_feature_descriptions.tsv.gz --n_resamples 0\n",
    "    \n",
    "We have included a small test ensemble (<u>*need to find a place to host this!!!*</u>). In this call we set `--n_resamples` to `0`. This means that you will not have to perform condition resampling for corems, which currently requires access to an SGE cluster. \n",
    "\n",
    "\n",
    "### STEP 3: Run col_resample on SGE cluster\n",
    "\n",
    "After loading all of the data from cMonkey2 into MongoDB and detecting corems, the `assembler.py` script will pause so that a computationally intensive procedure can be run on an SGE cluster. \n",
    "\n",
    "This step assigns conditions to the corems. In short, it computes brute force resamples of the gene expression data (default: 10,000 resamples) to determine the conditions in which genes from each corem are tightly co-expressed. For more information about this step, please refer to the [Supplementary Information of our 2014 paper](http://msb.embopress.org/highwire/filestream/49752/field_highwire_a_enclosures/0/supplementary-material.inline-supplementary-material-10.pdf?download=true). \n",
    "\n",
    "The script generates an output directory containing QSub scripts and a master control script called `resample.sh`. You should transfer this folder as well as the `resample.py` script located in the `assemble` folder of the `egrin2-tools` repository to the cluster to compute the resamples.\n",
    "\n",
    "When `assembler.py` pasues for this step, you will see the following message displayed on the screen. \n",
    "\n",
    ">Output Qsub scripts to *targetdir*\n",
    ">   \n",
    ">Transfer these documents to the cluster. Run 'resample.sh' with resample.py in your working directory to compute all resamples. \n",
    ">    \n",
    ">Once this is done, return here to finish processing corems.\n",
    ">   \n",
    ">Please type: 'Done' to continue\"\"\"\n",
    "\n",
    "Be sure that the SGE cluster on which the resampling is run has access to the MongoDB host, as the script directly writes the output of the resampling to this database. Once the resampling is finsihed (all resample database entries present), you can return to this prompt and type `Done` to finish the assembly\n",
    "\n",
    "**Alternatively**: if for some reason your session terminates prematurely or for any other reason there is an interruption at this step, you can start `assembler.py` with the flag `--finish_only True`. You must also include all of the original parameters to the `assembler.py` script. \n",
    "\n",
    "*Importantly, this still assumes that all of the information from the resampling procedure is contained in the MongoDB database.* \n",
    "\n",
    "For example, returning to our original `assembler.py` call, we would type:\n",
    "\n",
    "python assembler.py --organism eco --ratios ./ratios_eco_m3d.tsv.gz --targetdir ./ --ncbi_code 511145 --ensembledir ./eco-ens-m3d/ --col_annot ./E_coli_v4_Build_6.experiment_feature_descriptions.tsv.gz --n_resamples 0 --finish_only True\n",
    "\n",
    "To finish the assembly.\n",
    "\n",
    "### STEP 4: Finish assembly\n",
    "\n",
    "Having finished assembly, the `assembler.py` script will dump the MongoDB database to BSON, which are subsequently compressed in the supplied `--targetdir`. This file can be uncompressed and read directly into MongoDB using the [`mongorestore` command](http://docs.mongodb.org/manual/reference/program/mongorestore/), e.g.:\n",
    "\n",
    "    mongorestore eco_db\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
