{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div align=\"right\" vertical-align=\"middle\" style=\"border: 2px solid;border-radius: 5px;background-color:lightgrey;padding:5px;padding-right:20px;padding-left:10px;\">\n",
    "        <a style=\"color:black;text-decoration:none;\" title=\"Home\" href=\"../index.ipynb\">\n",
    "            <img src=\"../../css/iconmonstr-christmas-house-icon.svg\" height = \"30\" width = \"30\" style=\"display:inline\">\n",
    "        </a>\n",
    "        &nbsp;\n",
    "        <b>|</b>\n",
    "        &nbsp;\n",
    "        <a style=\"color:black;text-decoration:none;\" title=\"Build\" href=\"../build_docs/build.ipynb\">\n",
    "            <img src=\"../../css/iconmonstr-puzzle-icon.svg\" height = \"30\" width = \"30\" style=\"display:inline\">\n",
    "        </a>\n",
    "        <a style=\"color:black;text-decoration:none;\" title=\"Assemble\" href=\"assemble.ipynb\">\n",
    "            <img src=\"../../css/iconmonstr-puzzle-17-icon.svg\" height = \"30\" width = \"30\" style=\"display:inline\">\n",
    "        </a>\n",
    "        <a style=\"color:black;text-decoration:none;\" title=\"Query\" href=\"../query_docs/query.ipynb\">\n",
    "            <img src=\"../../css/iconmonstr-flask-3-icon.svg\" height = \"30\" width = \"30\" style=\"display:inline\">\n",
    "        </a>\n",
    "</div>\n",
    "\n",
    "# TUTORIAL: Assemble an EGRIN 2.0 Ensemble\n",
    "\n",
    "####*Important!!! This tutorial assumes you have access to a complete cMonkey2 ensemble.*\n",
    "\n",
    "### In a nutshell\n",
    "\n",
    "The `ASSEMBLE` scripts transfer and compile individual [cMonkey2](https://github.com/baliga-lab/cmonkey2) SQLite databases into an integrated MongoDB database. \n",
    "\n",
    "In addition, they perform several post-processing steps, including: detection of gene regulatory elements (GREs) by comparing individual bicluster motifs with [`TOMTOM`](http://meme.ebi.edu.au/meme/doc/tomtom.html) and clustering with [MCL](http://micans.org/mcl/), genome-wide scanning of motifs with FIMO, and detection of co-regulated modules or **corems** using link-community detection.\n",
    "\n",
    "### Requirements\n",
    "\n",
    "- MongoDB >= 2.4.9\n",
    "\n",
    "IMPORTANT: This tutorial currently assumes that `TOMTOM`, `MCL` and `FIMO` have already been run.\n",
    "\n",
    "A single GRE definition file is read from, eg:\n",
    "    \n",
    "    /ensemble-head-dir\n",
    "        /out.mot_metaclustering.txt.I45.txt\n",
    "        \n",
    "FIMO scans are read from each run sub-directory, eg:\n",
    "    \n",
    "    /ensemble-head-dir\n",
    "        /org-out-xxx\n",
    "            /fimo-outs\n",
    "                /fimo-out-xxxx.bz2\n",
    "\n",
    "Optional:\n",
    "- `row_annot`: tab-delimited row (gene) annotations. Will be downloaded from MicrobesOnline automatically using --ncbi_code if undefined \n",
    "- `col_annot`: tab-delimited column (condition) annotations.\n",
    "\n",
    "The format for these files will be described in detail below.\n",
    "\n",
    "Additionally, the Python modules described on the [Home page](../index.ipynb) are required to run these scripts.\n",
    "\n",
    "### Scripts\n",
    "\n",
    "- `assembler.py`: The control function for `ASSEMBLE` scripts.\n",
    "- `makeCorems.py`: Identifies corems\n",
    "- `resample_QSub.py`: Generates QSub script for submission of resamples to cluster\n",
    "- `sql2mongoDB.py`: Merges individual cMonkey SQLite dbs and post-processing data into MongoDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: assembler.py [-h] --organism ORGANISM --ratios RATIOS --targetdir\n",
      "                    TARGETDIR --ncbi_code NCBI_CODE [--cores CORES]\n",
      "                    [--ensembledir ENSEMBLEDIR] [--col_annot COL_ANNOT]\n",
      "                    [--host HOST] [--port PORT] [--prefix PREFIX]\n",
      "                    [--row_annot ROW_ANNOT]\n",
      "                    [--row_annot_matchCol ROW_ANNOT_MATCHCOL]\n",
      "                    [--gre2motif GRE2MOTIF] [--db DB]\n",
      "                    [--genome_annot GENOME_ANNOT]\n",
      "                    [--backbone_pval BACKBONE_PVAL]\n",
      "                    [--link_comm_score LINK_COMM_SCORE]\n",
      "                    [--link_comm_increment LINK_COMM_INCREMENT]\n",
      "                    [--link_comm_density_score LINK_COMM_DENSITY_SCORE]\n",
      "                    [--corem_size_threshold COREM_SIZE_THRESHOLD]\n",
      "                    [--n_resamples N_RESAMPLES] [--cluster CLUSTER]\n",
      "                    [--finish_only FINISH_ONLY] [--user USER]\n",
      "\n",
      "assemble.py - prepare cluster runs\n",
      "\n",
      "optional arguments:\n",
      "  -h, --help            show this help message and exit\n",
      "  --organism ORGANISM   3 letter organism code\n",
      "  --ratios RATIOS       Path to ratios file. Should be 'raw' (normalized)\n",
      "                        ratios, not the standardized ratios used by cMonkey\n",
      "  --targetdir TARGETDIR\n",
      "                        Storage path for MongoDB and corem data\n",
      "  --ncbi_code NCBI_CODE\n",
      "                        NCBI organism code\n",
      "  --cores CORES         Number local cores to use for corem C++ scripts\n",
      "  --ensembledir ENSEMBLEDIR\n",
      "                        Path to ensemble runs. Default: cwd\n",
      "  --col_annot COL_ANNOT\n",
      "                        Tab-delimited file with experiment annotations\n",
      "  --host HOST           MongoDB host. Default 'localhost'\n",
      "  --port PORT           MongoDB port\n",
      "  --prefix PREFIX       Ensemble run prefix. Default: *organism*-out-\n",
      "  --row_annot ROW_ANNOT\n",
      "                        Optional row (gene) annotation tab-delimited file. If\n",
      "                        not specified, annotations will be downloaded from\n",
      "                        MicrobesOnline using --ncbi_code.\n",
      "  --row_annot_matchCol ROW_ANNOT_MATCHCOL\n",
      "                        Name of column in row_annot that matches row names in\n",
      "                        ratios file.\n",
      "  --gre2motif GRE2MOTIF\n",
      "                        Motif->GRE clustering file\n",
      "  --db DB               Optional ensemble MongoDB database name\n",
      "  --genome_annot GENOME_ANNOT\n",
      "                        Optional genome annotation file. Automatically\n",
      "                        downloaded from MicrobesOnline using --ncbi_code\n",
      "  --backbone_pval BACKBONE_PVAL\n",
      "                        Significance pvalue for gene-gene backbone. Default =\n",
      "                        0.05.\n",
      "  --link_comm_score LINK_COMM_SCORE\n",
      "                        Scoring metric for link communities\n",
      "  --link_comm_increment LINK_COMM_INCREMENT\n",
      "                        Height increment for cutting agglomerative clustering\n",
      "                        of link communities\n",
      "  --link_comm_density_score LINK_COMM_DENSITY_SCORE\n",
      "                        Density score for evaluating link communities\n",
      "  --corem_size_threshold COREM_SIZE_THRESHOLD\n",
      "                        Defines minimum corem size. Default = 3.\n",
      "  --n_resamples N_RESAMPLES\n",
      "                        Number resamples to compute for corem condition\n",
      "                        assignment. Default = 10,000\n",
      "  --cluster CLUSTER     Run re-samples on cluster? Boolean.\n",
      "  --finish_only FINISH_ONLY\n",
      "                        Finish corems only. In case session gets dropped\n",
      "  --user USER           Cluster user name\n"
     ]
    }
   ],
   "source": [
    "%run ../../assemble/assembler.py -h"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### `ASSEMBLE` an EGRIN 2.0 ensemble\n",
    "\n",
    "In this tutorial we will see how you would `ASSEMBLE` an *Escherichia coli* EGRIN 2.0 ensemble using several example files and a couple of cMonkey2 runs, which we provide [here](./static/example_files/).\n",
    "\n",
    "### STEP 1: Generate optional input files\n",
    "\n",
    "First, let's explore the optional annotation files. Providing annotations for genes and conditions is a great way to enrich your analysis of the ensemble. You can get a better idea for the utility of these metainformation by following the [advanced mining tutorial](../query_docs/advanced_mining.ipynb)\n",
    "\n",
    "As noted above, the `row_annotations` file will be downloaded automatically from MicrobesOnline if a custom annotation is not provided. If you provide your own `row_annot` file, however, you will also need to specificy `--row_annot_matchCol`, which is the name of the column in your annotation file that matches the gene name used by cMonkey2 (i.e. the row names in your ratios file). \n",
    "\n",
    "The row annotation file should look like the annotation file supplied by MicrobesOnline, where each row specifies a gene and each of the columns specifies some information about that gene. Again, you must ensure that at least one of the columns contains gene names that match the gene names in the ratios file used by cMonkey2, in the case of MicrobesOnline, it is the `sysName` column.\n",
    "\n",
    "Here is an example annotation file for *E. coli* direct from MicrobesOnline, the file itself is available [here](./static/example_files/E_coli_v4_Build_6.experiment_feature_descriptions.tsv).\n",
    "\n",
    "<img src=\"./static/row_annot.png\" height = \"75%\" width = \"75%\" style=\"display:inline\">\n",
    "\n",
    "The `col_annot` file provides metainformation about. \n",
    "\n",
    "\n",
    "\n",
    "<img src=\"./static/experiment_annotation.png\" height = \"75%\" width = \"75%\" style=\"display:inline\">\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
