
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Assemble Tutorial &#8212; egrin2-tools 1.0.0 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.0.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Query" href="query.html" />
    <link rel="prev" title="Assemble" href="assemble.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head>
  <body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="assemble-tutorial">
<h1>Assemble Tutorial<a class="headerlink" href="#assemble-tutorial" title="Permalink to this headline">¶</a></h1>
<p><strong>Important!!! This tutorial assumes you have access to a complete cMonkey2 ensemble.</strong></p>
<div class="section" id="in-a-nutshell">
<h2>In a nutshell<a class="headerlink" href="#in-a-nutshell" title="Permalink to this headline">¶</a></h2>
<p>The ASSEMBLE scripts transfer and compile individual cMonkey2 SQLite databases into an integrated MongoDB database.</p>
<p>In addition, they perform several post-processing steps, including: detection of gene regulatory elements (GREs) by comparing individual bicluster motifs with TOMTOM and clustering with MCL, genome-wide scanning of motifs with FIMO, and detection of co-regulated modules or <strong>corems</strong> using link-community detection.</p>
</div>
<div class="section" id="requirements">
<h2>Requirements<a class="headerlink" href="#requirements" title="Permalink to this headline">¶</a></h2>
<blockquote>
<div><ul class="simple">
<li>MongoDB &gt;= 2.4.9</li>
<li>compiled C++ scripts for corem detection, available here</li>
</ul>
</div></blockquote>
<p>IMPORTANT: This tutorial currently assumes that <code class="docutils literal"><span class="pre">TOMTOM</span></code>, <code class="docutils literal"><span class="pre">MCL</span></code> and <code class="docutils literal"><span class="pre">FIMO</span></code> have already been run.</p>
<p>A single GRE definition file is read from, eg:</p>
<div class="highlight-none"><div class="highlight"><pre><span></span>/ensemble-head-dir
    /out.mot_metaclustering.txt.I45.txt
</pre></div>
</div>
<p>FIMO scans are read from each run sub-directory, eg:</p>
<div class="highlight-none"><div class="highlight"><pre><span></span>/ensemble-head-dir
    /org-out-xxx
        /fimo-outs
            /fimo-out-xxxx.bz2
</pre></div>
</div>
<p>Optional:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">row_annot</span></code>: tab-delimited row (gene) annotations. Will be downloaded from MicrobesOnline automatically using <code class="docutils literal"><span class="pre">--ncbi_code</span></code> if undefined</li>
<li><code class="docutils literal"><span class="pre">col_annot</span></code>: tab-delimited column (condition) annotations.</li>
</ul>
</div></blockquote>
<p>The format for these files will be described in detail below.</p>
<p>Additionally, the Python modules described on the Home page are required to run these scripts.</p>
</div>
<div class="section" id="scripts">
<h2>Scripts<a class="headerlink" href="#scripts" title="Permalink to this headline">¶</a></h2>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">assembler.py</span></code>: The control function for ASSEMBLE scripts.</li>
<li><code class="docutils literal"><span class="pre">makeCorems.py</span></code>: Identifies corems using C++ scripts compiled above</li>
<li><code class="docutils literal"><span class="pre">resample_QSub.py</span></code>: Generates QSub script for submission of resamples to cluster</li>
<li><code class="docutils literal"><span class="pre">sql2mongoDB.py</span></code>: Merges individual cMonkey SQLite dbs and post-processing data into MongoDB</li>
</ul>
</div></blockquote>
<div class="highlight-none"><div class="highlight"><pre><span></span>%run ..//assembler.py -h

usage: assembler.py [-h] --organism ORGANISM --ratios RATIOS --targetdir TARGETDIR --ncbi_code NCBI_CODE [--cores CORES] [--ensembledir ENSEMBLEDIR] [--col_annot COL_ANNOT] [--host HOST] [--port PORT] [--prefix PREFIX] [--row_annot ROW_ANNOT] [--row_annot_matchCol ROW_ANNOT_MATCHCOL] [--gre2motif GRE2MOTIF] [--db DB] [--genome_annot GENOME_ANNOT] [--backbone_pval BACKBONE_PVAL] [--link_comm_score LINK_COMM_SCORE] [--link_comm_increment LINK_COMM_INCREMENT] [--link_comm_density_score LINK_COMM_DENSITY_SCORE] [--corem_size_threshold COREM_SIZE_THRESHOLD] [--n_resamples N_RESAMPLES] [--cluster CLUSTER] [--finish_only FINISH_ONLY] [--user USER]

assemble.py - prepare cluster runs

optional arguments:
  -h, --help            show this help message and exit
  --organism ORGANISM   3 letter organism code
  --ratios RATIOS       Path to ratios file. Should be &#39;raw&#39; (normalized) ratios, not the standardized ratios used by cMonkey
  --targetdir TARGETDIR Storage path for MongoDB and corem data
  --ncbi_code NCBI_CODE NCBI organism code
  --cores CORES Number  local cores to use for corem C++ scripts
  --ensembledir ENSEMBLEDIR
                        Path to ensemble runs. Default: cwd
  --col_annot COL_ANNOT Tab-delimited file with experiment annotations
  --host HOST           MongoDB host. Default &#39;localhost&#39;
  --port PORT           MongoDB port
  --prefix PREFIX       Ensemble run prefix. Default: organism-out-
  --row_annot ROW_ANNOT Optional row (gene) annotation tab-delimited file. If not specified, annotations will be downloaded from MicrobesOnline using --ncbi_code.
  --row_annot_matchCol ROW_ANNOT_MATCHCOL
                        Name of column in row_annot that matches row names in ratios file.
  --gre2motif GRE2MOTIF Motif-&gt;GRE clustering file
  --db DB               Optional ensemble MongoDB database name
  --genome_annot GENOME_ANNOT
                        Optional genome annotation file. Automatically downloaded from MicrobesOnline using --ncbi_code
  --backbone_pval BACKBONE_PVAL
                        Significance pvalue for gene-gene backbone. Default = 0.05.
  --link_comm_score LINK_COMM_SCORE
                        Scoring metric for link communities
  --link_comm_increment LINK_COMM_INCREMENT
                        Height increment for cutting agglomerative clustering of link communities
  --link_comm_density_score LINK_COMM_DENSITY_SCORE
                        Density score for evaluating link communities
  --corem_size_threshold COREM_SIZE_THRESHOLD Defines minimum corem size. Default = 3.
  --n_resamples N_RESAMPLES
                        Number resamples to compute for corem condition assignment. Default = 10,000
  --cluster CLUSTER     Run re-samples on cluster? Boolean.
  --finish_only FINISH_ONLY
                        Finish corems only. In case session gets dropped
  --user USER Cluster   user name
</pre></div>
</div>
</div>
<div class="section" id="assemble-an-egrin-2-0-ensemble">
<h2>ASSEMBLE an EGRIN 2.0 ensemble<a class="headerlink" href="#assemble-an-egrin-2-0-ensemble" title="Permalink to this headline">¶</a></h2>
<p>In this tutorial we will see how you would ASSEMBLE an <em>Escherichia coli</em> EGRIN 2.0 ensemble using several example files and a couple of cMonkey2 runs, which we provide here.</p>
</div>
<div class="section" id="step-1-generate-optional-input-files">
<h2>STEP 1: Generate optional input files<a class="headerlink" href="#step-1-generate-optional-input-files" title="Permalink to this headline">¶</a></h2>
<p>First, let’s explore the optional annotation files. Providing annotations for genes and conditions is a great way to enrich your analysis of the ensemble. You can get a better idea for the utility of these metainformation by following the advanced mining tutorial</p>
<div class="section" id="row-annot">
<h3><code class="docutils literal"><span class="pre">row_annot</span></code><a class="headerlink" href="#row-annot" title="Permalink to this headline">¶</a></h3>
<p>As noted above, the <code class="docutils literal"><span class="pre">row_annot</span></code> file will be downloaded automatically from MicrobesOnline if a custom annotation is not provided. If you provide your own row_annot file, however, you will also need to specificy <code class="docutils literal"><span class="pre">--row_annot_matchCol</span></code>, which is the name of the column in your annotation file that matches the gene name used by cMonkey2 (i.e. the row names in your ratios file).</p>
<p>The row annotation file should look like the annotation file supplied by MicrobesOnline, where each row specifies a gene and each of the columns specifies some information about that gene. Again, you must ensure that at least one of the columns contains gene names that match the gene names in the ratios file used by cMonkey2, in the case of MicrobesOnline, it is the <code class="docutils literal"><span class="pre">sysName</span></code> column.</p>
<p>Here is an example annotation file for <em>E. coli</em> direct from MicrobesOnline, the file itself is available here.</p>
<div class="figure">
<img alt="Example row_annot file" src="_images/row_annot.png" />
</div>
</div>
<div class="section" id="col-annot">
<h3><code class="docutils literal"><span class="pre">col_annot</span></code><a class="headerlink" href="#col-annot" title="Permalink to this headline">¶</a></h3>
<p>The col_annot file provides metainformation about each experiment. Like the row_annot file, these annotations are optional, but they can be valuable for making sense of ensemble predicitions.</p>
<p>Please note that the file format is different here. Each row contains a particular experimental meta-annotation followed by several required descriptions: (1) experiment_name, (2) feature_name, (3) value, (4) feature_units, (5) feature_type. The experiment_name column should match the experiment name in the ratios file.</p>
<p>The <cite>col_annot</cite> file should look like the tab-delimited file depicted below. You can download an <em>E. coli</em> <cite>col_annot</cite> file to use as a template here</p>
<div class="figure">
<img alt="Example col_annot file" src="_images/col_annot.png" />
</div>
</div>
</div>
<div class="section" id="step-2-run-assembler-py">
<h2>STEP 2: Run <code class="docutils literal"><span class="pre">assembler.py</span></code><a class="headerlink" href="#step-2-run-assembler-py" title="Permalink to this headline">¶</a></h2>
<p><strong>Important!!! TOMTOM, MCL, and FIMO should be run prior to assembly. Otherwise the ensemble will not contain GREs or motif scans.</strong></p>
<p>To run the assembler, you must supply several files as well as specify where you would like to host the MongoDB database. If you do not supply a host for the MongoDB databse, it will be stored on <code class="docutils literal"><span class="pre">localhost</span></code>. At a minimum, you should supply:</p>
<blockquote>
<div><ul class="simple">
<li><code class="docutils literal"><span class="pre">--organism</span></code>: 3 letter organism code</li>
<li><code class="docutils literal"><span class="pre">--ratios</span></code>: Path to ratios file. Should be ‘raw’ (normalized) ratios, not the standardized ratios used by cMonkey</li>
<li><code class="docutils literal"><span class="pre">--targetdir</span></code>: Storage path for MongoDB and corem data</li>
<li><code class="docutils literal"><span class="pre">--ncbi_code</span></code>: NCBI organism code</li>
<li><code class="docutils literal"><span class="pre">--ensembledir</span></code>: Path to ensemble runs. Default: cwd</li>
</ul>
</div></blockquote>
<p>Assuming your terminal current working directory is within the E. coli ensemble, with the cMonkey2 runs located in the subfolder <code class="docutils literal"><span class="pre">./eco-ens-m3d/</span></code>, you can call the assembler as follows:</p>
<div class="highlight-none"><div class="highlight"><pre><span></span>python assembler.py --organism eco --ratios ./ratios_eco_m3d.tsv.gz --targetdir ./ --ncbi_code 511145 --ensembledir ./eco-ens-m3d/ --col_annot ./E_coli_v4_Build_6.experiment_feature_descriptions.tsv.gz --n_resamples 0
</pre></div>
</div>
<p>We have included a small test ensemble (need to find a place to host this!!!). In this call we set <code class="docutils literal"><span class="pre">--n_resamples</span></code> to 0. This means that you will not have to perform condition resampling for corems, which currently requires access to an SGE cluster.</p>
<p>The MongoDB Database Schema is available here</p>
</div>
<div class="section" id="step-3-run-col-resample-on-sge-cluster">
<h2>STEP 3: Run <code class="docutils literal"><span class="pre">col_resample</span></code> on SGE cluster<a class="headerlink" href="#step-3-run-col-resample-on-sge-cluster" title="Permalink to this headline">¶</a></h2>
<p>After loading all of the data from cMonkey2 into MongoDB and detecting corems, the <code class="docutils literal"><span class="pre">assembler.py</span></code> script will pause so that a computationally intensive procedure can be run on an SGE cluster.</p>
<p>This step assigns conditions to the corems. In short, it computes brute force resamples of the gene expression data (default: 10,000 resamples) to determine the conditions in which genes from each corem are tightly co-expressed. For more information about this step, please refer to the <a class="reference external" href="http://msb.embopress.org/highwire/filestream/49752/field_highwire_a_enclosures/0/supplementary-material.inline-supplementary-material-10.pdf?download=true">Supplementary Information of our 2014 paper</a>.</p>
<p>The script generates an output directory containing QSub scripts and a master control script called <code class="docutils literal"><span class="pre">resample.sh</span></code>. You should transfer this folder as well as the <code class="docutils literal"><span class="pre">resample.py</span></code> script located in the assemble folder of the egrin2-tools repository to the cluster to compute the resamples.</p>
<p>When <code class="docutils literal"><span class="pre">assembler.py</span></code> pauses for this step, you will see the following message displayed on the screen.</p>
<div class="highlight-none"><div class="highlight"><pre><span></span>Output Qsub scripts to targetdir

Transfer these documents to the cluster. Run &#39;resample.sh&#39; with resample.py in your working directory to compute all resamples.

Once this is done, return here to finish processing corems.

Please type: &#39;Done&#39; to continue
</pre></div>
</div>
<p>Be sure that the SGE cluster on which the resampling is run has access to the MongoDB host, as the script directly writes the output of the resampling to this database. Once the resampling is finsihed (all resample database entries present), you can return to this prompt and type Done to finish the assembly</p>
<p><strong>Alternatively:</strong> if for some reason your session terminates prematurely or for any other reason there is an interruption at this step, you can start <code class="docutils literal"><span class="pre">assembler.py</span></code> with the flag <code class="docutils literal"><span class="pre">--finish_only</span> <span class="pre">True</span></code>. You must also include all of the original parameters to the <code class="docutils literal"><span class="pre">assembler.py</span></code> script.</p>
<p>Importantly, this still assumes that all of the information from the resampling procedure is contained in the MongoDB database.</p>
<p>For example, returning to our original <code class="docutils literal"><span class="pre">assembler.py</span></code> call, we would type:</p>
<div class="highlight-none"><div class="highlight"><pre><span></span>python assembler.py --organism eco --ratios ./ratios_eco_m3d.tsv.gz --targetdir ./ --ncbi_code 511145 --ensembledir ./eco-ens-m3d/ --col_annot ./E_coli_v4_Build_6.experiment_feature_descriptions.tsv.gz --n_resamples 0 --finish_only True
</pre></div>
</div>
<p>To finish the assembly.</p>
</div>
<div class="section" id="step-4-finish-assembly">
<h2>STEP 4: Finish assembly<a class="headerlink" href="#step-4-finish-assembly" title="Permalink to this headline">¶</a></h2>
<p>Having finished assembly, the <code class="docutils literal"><span class="pre">assembler.py</span></code> script will dump the MongoDB database to BSON, which are subsequently compressed in the supplied <code class="docutils literal"><span class="pre">--targetdir</span></code>. This file can be uncompressed and read directly into MongoDB using the <code class="docutils literal"><span class="pre">mongorestore</span></code> command, e.g.:</p>
<div class="highlight-none"><div class="highlight"><pre><span></span>mongorestore eco_db
</pre></div>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<p class="logo">
  <a href="index.html">
    <img class="logo" src="_static/egrin2_logo_80px.png" alt="Logo"/>
    
    <h1 class="logo logo-name">egrin2-tools</h1>
    
  </a>
</p>



<p class="blurb">EGRIN2 tool chain</p>




<p>
<iframe src="https://ghbtns.com/github-btn.html?user=baliga-lab&repo=egrin2-tools&type=watch&count=true&size=large&v=2"
  allowtransparency="true" frameborder="0" scrolling="0" width="200px" height="35px"></iframe>
</p>





<h3>Navigation</h3>
<p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="introduction.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="build.html">Build</a></li>
<li class="toctree-l1"><a class="reference internal" href="build_tutorial.html">Build Tutorial</a></li>
<li class="toctree-l1"><a class="reference internal" href="assemble.html">Assemble</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Assemble Tutorial</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#in-a-nutshell">In a nutshell</a></li>
<li class="toctree-l2"><a class="reference internal" href="#requirements">Requirements</a></li>
<li class="toctree-l2"><a class="reference internal" href="#scripts">Scripts</a></li>
<li class="toctree-l2"><a class="reference internal" href="#assemble-an-egrin-2-0-ensemble">ASSEMBLE an EGRIN 2.0 ensemble</a></li>
<li class="toctree-l2"><a class="reference internal" href="#step-1-generate-optional-input-files">STEP 1: Generate optional input files</a></li>
<li class="toctree-l2"><a class="reference internal" href="#step-2-run-assembler-py">STEP 2: Run <code class="docutils literal"><span class="pre">assembler.py</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#step-3-run-col-resample-on-sge-cluster">STEP 3: Run <code class="docutils literal"><span class="pre">col_resample</span></code> on SGE cluster</a></li>
<li class="toctree-l2"><a class="reference internal" href="#step-4-finish-assembly">STEP 4: Finish assembly</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="query.html">Query</a></li>
<li class="toctree-l1"><a class="reference internal" href="basic_mining_agglom.html">Basic Mining Using agglom</a></li>
<li class="toctree-l1"><a class="reference internal" href="basic_mining_coremfinder.html">Basic Mining Using coremFinder</a></li>
<li class="toctree-l1"><a class="reference internal" href="advanced_mining.html">Advanced Mining</a></li>
<li class="toctree-l1"><a class="reference internal" href="plot_expression_tutorial.html">Plot Expression Tutorial</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="assemble.html" title="previous chapter">Assemble</a></li>
      <li>Next: <a href="query.html" title="next chapter">Query</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2017, Aaron Brooks, David Reiss, Wei-ju Wu, Micheleen Harris.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.6.4</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.10</a>
      
      |
      <a href="_sources/assemble_tutorial.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    
    <a href="https://github.com/baliga-lab/egrin2-tools" class="github">
        <img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png" alt="Fork me on GitHub"  class="github"/>
    </a>
    

    
  </body>
</html>